Add style and layout interleaving for `anchor()`

RWDavid opened the pull request:
#### 5f0104f4e2cd965e5622fb88090d7425bc464801
<pre>
Add style and layout interleaving for `anchor()`
<a href="https://bugs.webkit.org/show_bug.cgi?id=275938">https://bugs.webkit.org/show_bug.cgi?id=275938</a>
<a href="https://rdar.apple.com/130642534">rdar://130642534</a>

Reviewed by Antti Koivisto.

This patch introduces an initial implementation of style and layout
interleaving for anchor positioning. For each anchor-positioned element,
we perform a layout to get its containing block, then bounce back to
style resolution to determine its anchors, and then finally attempt
to resolve the element with another style resolution.

The actual resolution of `anchor()` is not implemented yet, (this patch
is a precursor for its implementation). As a result, there are no
behavioral changes.

This patch also is not fully spec-compliant when finding anchors for
each anchor-positioned element. It only attempts to find the last
anchor element in tree order with a matching anchor name. Further
patches will build on this patch to correctly determine anchors and
actually resolve anchor-positioned elements.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSAnchorValue.cpp:
(WebCore::CSSAnchorValue::collectAnchorNames const):
* Source/WebCore/css/CSSAnchorValue.h:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeAnchor):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resolveStyle):
(WebCore::Document::isInStyleInterleavedLayoutForSelfOrAncestor const):
* Source/WebCore/dom/Document.h:
(WebCore::Document::isInStyleInterleavedLayout const):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::invalidateAncestorsForAnchor):
(WebCore::Element::invalidateForResumingAnchorPositionedElementResolution):
* Source/WebCore/dom/Element.h:
* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
(WebCore::LocalFrameViewLayoutContext::performLayout):
(WebCore::LocalFrameViewLayoutContext::runOrScheduleAsynchronousTasks):
* Source/WebCore/style/AnchorPositionEvaluator.cpp: Added.
(WebCore::Style::AnchorPositionEvaluator::resolveAnchorValue):
* Source/WebCore/style/AnchorPositionEvaluator.h: Added.
* Source/WebCore/style/MatchedDeclarationsCache.cpp:
(WebCore::Style::MatchedDeclarationsCache::isCacheable):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertLength):
* Source/WebCore/style/StyleResolver.cpp:
* Source/WebCore/style/StyleScope.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::popParent):
(WebCore::Style::TreeResolver::resolveComposedTree):
(WebCore::Style::TreeResolver::resolve):
(WebCore::Style::TreeResolver::updateAnchorPositioningState):
(WebCore::Style::TreeResolver::findAnchorsForAnchorPositionedElement):
(WebCore::Style::elementIsInContainingBlockChain):
(WebCore::Style::TreeResolver::findLastAcceptableAnchorWithName):
(WebCore::Style::TreeResolver::updateAnchorPositioningStateInInitialContainingBlock):
* Source/WebCore/style/StyleTreeResolver.h:
(WebCore::Style::TreeResolver::hasUnresolvedAnchorPositionedElements const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:

Canonical link: <a href="https://commits.webkit.org/280543@main">https://commits.webkit.org/280543@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/91748303d649265f705afaa17b528aea56a94115

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56915 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36243 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9389 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60535 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7358 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59043 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43866 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7548 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46100 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5168 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58945 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34041 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49120 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26958 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30821 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6451 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6363 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52781 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6722 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62216 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/828 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6828 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53357 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/831 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49175 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53401 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12580 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/702 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32072 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33157 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34242 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32903 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
```diff
@@ -283,8 +283,8 @@ void LocalFrameViewLayoutContext::runOrScheduleAsynchronousTasks()
 
-    if (document()->isResolvingContainerQueries()) {
-        // We are doing layout from style resolution to resolve container queries.
+    if (document()->isResolvingContainerQueries() || document()->isResolvingAnchorPositionedElements()) {
```
We could come up with a single name for this (`isInLayoutInterleavedStyleResolution()` or something?). It is tested in a few places.

anttijk commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -0,0 +1,55 @@
+    if (anchorPositionedElementStateIt == m_anchorPositionedElementStates.end())
+        anchorPositionedElementState = &m_anchorPositionedElementStates.add(element, AnchorPositionedElementState { }).iterator->value;
+    else
+        anchorPositionedElementState = &anchorPositionedElementStateIt->value;
```
This can be written as

```
auto& anchorPositionedElementState = m_anchorPositionedElementStates.ensure(element, [&] {
    return AnchorPositionedElementState { };
}).iterator->value;
```
or thereabouts, saving a hash lookup and looking nicer.

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -874,6 +877,23 @@ void TreeResolver::popParent()
+    if (m_nextAnchorPositionedElementCanFindAnchors) {
+        auto it = m_unresolvedAnchorPositionedElementsForContainingBlock.find(parentElement);
+        if (it != m_unresolvedAnchorPositionedElementsForContainingBlock.end()) {
+            if (!it->value.isEmpty()) {
```
I think you can just use `m_unresolvedAnchorPositionedElementsForContainingBlock.get(parentElement)`. You'll get an empty Vector the key doesn't exist which is what you want here.

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1265,5 +1317,129 @@ bool postResolutionCallbacksAreSuspended()
+                if (anchorNameMapIt == m_anchorNameMap.end())
+                    m_anchorNameMap.add(anchorName, Vector<Ref<Element>> { element });
+                else
+                    anchorNameMapIt->value.append(element);
```
similarly, you can use `ensure()` here for more compact and efficient code.

anttijk commented on Source/WebCore/style/AnchorPositionEvaluator.h:
```diff
@@ -0,0 +1,59 @@
+    WeakHashMap<Element, Style::AnchorPositionedElementState, WeakPtrImplWithEventTargetData>& anchorPositionedElementStates();
+
+private:
+    WeakHashMap<Element, Style::AnchorPositionedElementState, WeakPtrImplWithEventTargetData> m_anchorPositionedElementStates;
```
You have `m_anchorPositionedElementStates` maps in two places, here in AnchorPositionEvaluator and also in TreeResolver. I suppose things flow between these maps? I wonder if the names could convey some difference.

RWDavid replied:
They actually refer to the same map (note the reference in TreeResolver). But I will adjust the ownership of the map such that this does not occur.

anttijk commented on Source/WebCore/style/AnchorPositionEvaluator.h:
```diff
@@ -0,0 +1,59 @@
+
+class AnchorPositionEvaluator {
+    WTF_MAKE_FAST_ALLOCATED;
+    WTF_MAKE_NONCOPYABLE(AnchorPositionEvaluator);
```
Types named "Evaluator" are generally stack-allocated transient helpers. Since this is persistent (owned by Document) and owns things (the anchor position state map) it should probably have a different name, `DocumentAnchorPositioningState` or something.

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1265,5 +1317,129 @@ bool postResolutionCallbacksAreSuspended()
 
+void TreeResolver::updateAnchorPositioningState(Element& element, const RenderStyle* style, bool& shouldSkipChildrenForAnchorPositionedElement)
+{
+    shouldSkipChildrenForAnchorPositionedElement = false;
```
It would be nicer to use enum class function return value instead of returning a boolean via a reference.

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1265,5 +1317,129 @@ bool postResolutionCallbacksAreSuspended()
+
+        } else {
+            // At this point, we should have layout information.
+            if (!anchorPositionedElementState.hasBeenResolved) {
```
you could restructure this function using early return style (which WebKit often uses) so it doesn't end being a tower of ifs.

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1265,5 +1317,129 @@ bool postResolutionCallbacksAreSuspended()
+                    if (it == m_unresolvedAnchorPositionedElementsForContainingBlock.end())
+                        m_unresolvedAnchorPositionedElementsForContainingBlock.add(*containingBlock, Vector<Ref<Element>> { element });
+                    else
+                        it->value.append(element);
```
similarly ensure()

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1265,5 +1317,129 @@ bool postResolutionCallbacksAreSuspended()
+        return { };
+
+    auto& anchors = it->value;
+    for (auto anchorIt = anchors.rbegin(); anchorIt != anchors.rend(); ++anchorIt) {
```
you can use modern loop

```
for (auto& anchor : makeReversedRange(anchors)) {
```

anttijk commented on Source/WebCore/style/StyleTreeResolver.h:
```diff
@@ -150,6 +158,15 @@ class TreeResolver {
 
+    AnchorPositionEvaluator& m_anchorPositionEvaluator;
+    HashSet<Ref<Element>> m_anchorElements;
+    WeakHashMap<Element, AnchorPositionedElementState, WeakPtrImplWithEventTargetData>& m_anchorPositionedElementStates;
```
You can use regular HashMap in TreeResolver since it is a transient type.

AnchorPositionedElementState is a rather complex struct (with maps as members) so it might be better to make this `HashMap<Ref<Element>, std::unique_ptr<AnchorPositionedElementState>>`. This will also make some code look nicer since you can use `get()` instead of `find()`. This goes for the other similar map too.

anttijk commented on Source/WebCore/style/StyleTreeResolver.h:
```diff
@@ -150,6 +158,15 @@ class TreeResolver {
+    Vector<Ref<Element>> m_unresolvedAnchorPositionedElementsForInitialContainingBlock;
+    HashMap<String, Vector<Ref<Element>>> m_anchorNameMap;
+    bool m_hasUnresolvedAnchorPositionedElements { false };
+    bool m_nextAnchorPositionedElementCanFindAnchors { false };
```
Wonder if some of these bools are equivalent to testing some of the vectors/maps for isEmpty()?

Something like `m_canFindAnchorForNextAnchorPositionedElement()` would be a more canonical name (we try to name boolean starting with a true/false verb.

RWDavid replied:
Both booleans are a little bit more involved, so there is not a simple way of querying the existing data structures to reconstruct either boolean.

I will use your suggested name.

anttijk reviewed (changes requested):
looks pretty good!

anttijk commented on Source/WebCore/dom/Document.h:
```diff
@@ -2620,6 +2626,7 @@ class Document
 
     mutable std::unique_ptr<CSSParserContext> m_cachedCSSParserContext;
     mutable std::unique_ptr<PermissionsPolicy> m_permissionsPolicy;
+    mutable std::unique_ptr<Style::AnchorPositionEvaluator> m_anchorPositionEvaluator;
```
This could go to `Style::Scope` where m_queryContainerStates are. 

It is slightly awkward since they are only used for the top-level scope, not shadow tree scopes. Might still be good for consistency and keeping Document from getting even more unwieldy (Style::Scope should be split into top-level scope class and shadow level scope class at some point).

mdubet commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1265,5 +1312,125 @@ bool postResolutionCallbacksAreSuspended()
+}
+
+// Precondition: containingBlock is nullptr if and only if containingBlock is the initial containing block.
+auto TreeResolver::findLastAcceptableAnchorWithName(String anchorName, const Element* containingBlock) -> std::optional<Ref<Element>>
```
There is no reason to use the `auto` return type syntax here ?

anttijk commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
```diff
@@ -39,6 +39,7 @@
 #import <WebCore/FloatRect.h>
 #import <WebCore/IntRect.h>
 #import <WebKit/WKMenuItemIdentifiersPrivate.h>
+#import <WebKit/WKWebViewConfigurationPrivate.h>
```
stray change or build fix?

RWDavid replied:
it is a build fix

RWDavid committed:
Add style and layout interleaving for `anchor()`
https://bugs.webkit.org/show_bug.cgi?id=275938
rdar://130642534

Reviewed by Antti Koivisto.

This patch introduces an initial implementation of style and layout
interleaving for anchor positioning. For each anchor-positioned element,
we perform a layout to get its containing block, then bounce back to
style resolution to determine its anchors, and then finally attempt
to resolve the element with another style resolution.

The actual resolution of `anchor()` is not implemented yet, (this patch
is a precursor for its implementation). As a result, there are no
behavioral changes.

This patch also is not fully spec-compliant when finding anchors for
each anchor-positioned element. It only attempts to find the last
anchor element in tree order with a matching anchor name. Further
patches will build on this patch to correctly determine anchors and
actually resolve anchor-positioned elements.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSAnchorValue.cpp:
(WebCore::CSSAnchorValue::collectAnchorNames const):
* Source/WebCore/css/CSSAnchorValue.h:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeAnchor):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resolveStyle):
(WebCore::Document::isInStyleInterleavedLayoutForSelfOrAncestor const):
* Source/WebCore/dom/Document.h:
(WebCore::Document::isInStyleInterleavedLayout const):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::invalidateAncestorsForAnchor):
(WebCore::Element::invalidateForResumingAnchorPositionedElementResolution):
* Source/WebCore/dom/Element.h:
* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
(WebCore::LocalFrameViewLayoutContext::performLayout):
(WebCore::LocalFrameViewLayoutContext::runOrScheduleAsynchronousTasks):
* Source/WebCore/style/AnchorPositionEvaluator.cpp: Added.
(WebCore::Style::AnchorPositionEvaluator::resolveAnchorValue):
* Source/WebCore/style/AnchorPositionEvaluator.h: Added.
* Source/WebCore/style/MatchedDeclarationsCache.cpp:
(WebCore::Style::MatchedDeclarationsCache::isCacheable):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertLength):
* Source/WebCore/style/StyleResolver.cpp:
* Source/WebCore/style/StyleScope.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::popParent):
(WebCore::Style::TreeResolver::resolveComposedTree):
(WebCore::Style::TreeResolver::resolve):
(WebCore::Style::TreeResolver::updateAnchorPositioningState):
(WebCore::Style::TreeResolver::findAnchorsForAnchorPositionedElement):
(WebCore::Style::elementIsInContainingBlockChain):
(WebCore::Style::TreeResolver::findLastAcceptableAnchorWithName):
(WebCore::Style::TreeResolver::updateAnchorPositioningStateInInitialContainingBlock):
* Source/WebCore/style/StyleTreeResolver.h:
(WebCore::Style::TreeResolver::hasUnresolvedAnchorPositionedElements const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:

Canonical link: https://commits.webkit.org/280543@main
